fw/rbcx-coprocessor/src/MotorController.cpp

```diff
@@ -0,0 +1,187 @@
+#include "MotorController.hpp"
+#include "Bsp.hpp"
+#include "Motor.hpp"
+#include "utils/Debug.hpp"
+#include "utils/TaskWrapper.hpp"
+
+#include "semphr.h"
+#include "stm32f1xx_ll_tim.h"
+#include <climits>
+#include <cstdlib>
+
+static void setPwmValue(TIM_TypeDef* timer, uint8_t motorIndex, uint16_t value);
+static void setMotorPower(uint8_t motorIndex, int32_t power, bool brake);
+
+static constexpr uint16_t maxPwm = 2000;
+static std::array<Motor, 4> motor;
+static TaskWrapper<1024> motorTask;
+static void taskFunc();
+static SemaphoreHandle_t motorSem;
+static StaticSemaphore_t motorSemStruct;
+
+void motorInit() {
+    LL_TIM_InitTypeDef pwmInit;
+    LL_TIM_StructInit(&pwmInit);
+
+    pwmInit.Prescaler = 0;
+    // this sets interrupts flag when counter reachs TOP:
+    pwmInit.CounterMode = LL_TIM_COUNTERMODE_CENTER_DOWN;
+    pwmInit.Autoreload = maxPwm;
+    pwmInit.ClockDivision = LL_TIM_CLOCKDIVISION_DIV1;
+    pwmInit.RepetitionCounter = 0;
+
+    LL_TIM_OC_InitTypeDef ocInit;
+    LL_TIM_OC_StructInit(&ocInit);
+    ocInit.OCMode = LL_TIM_OCMODE_PWM2;
+    ocInit.OCState = LL_TIM_OCSTATE_ENABLE;
+
+    ocInit.OCNState = LL_TIM_OCSTATE_ENABLE;
+    ocInit.CompareValue = 0;
+    ocInit.OCPolarity = LL_TIM_OCPOLARITY_HIGH;
+    ocInit.OCNPolarity = LL_TIM_OCPOLARITY_HIGH;
+    ocInit.OCIdleState = LL_TIM_OCIDLESTATE_HIGH;
+    ocInit.OCNIdleState = LL_TIM_OCIDLESTATE_HIGH;
+
+    LL_TIM_Init(pwmTimer, &pwmInit);
+    for (uint16_t channel = LL_TIM_CHANNEL_CH1; channel != 0; channel <<= 4) {
+        LL_TIM_OC_Init(pwmTimer, channel, &ocInit);
+        LL_TIM_OC_EnablePreload(pwmTimer, channel);
+    }
+
+    LL_TIM_SetOffStates(pwmTimer, LL_TIM_OSSI_DISABLE, LL_TIM_OSSR_ENABLE);
+    LL_TIM_GenerateEvent_UPDATE(pwmTimer);
+    LL_TIM_EnableAllOutputs(pwmTimer);
+    LL_TIM_EnableCounter(pwmTimer);
+
+    for (int motorIndex : { 0, 1, 2, 3 }) {
+        setMotorPower(motorIndex, 0, false);
+    }
+
+    LL_TIM_ENCODER_InitTypeDef encInit;
+    encInit.EncoderMode = LL_TIM_ENCODERMODE_X4_TI12;
+    encInit.IC1Polarity = LL_TIM_IC_POLARITY_RISING;
+    encInit.IC1ActiveInput = LL_TIM_ACTIVEINPUT_DIRECTTI;
+    encInit.IC1Prescaler = LL_TIM_ICPSC_DIV1;
+    encInit.IC1Filter = LL_TIM_IC_FILTER_FDIV1;
+    encInit.IC2Polarity = LL_TIM_IC_POLARITY_RISING;
+    encInit.IC2ActiveInput = LL_TIM_ACTIVEINPUT_DIRECTTI;
+    encInit.IC2Prescaler = LL_TIM_ICPSC_DIV1;
+    encInit.IC2Filter = LL_TIM_IC_FILTER_FDIV1;
+    for (auto timer : encoderTimer) {
+        LL_TIM_ENCODER_Init(timer, &encInit);
+        LL_TIM_EnableCounter(timer);
+    }
+
+    motorSem = xSemaphoreCreateBinaryStatic(&motorSemStruct);
```

Comment:
IMHO na tohle chceš spíš mutex než binary semaphore, protože je to "mutual exclusion" a ne synchronizace mezi ISR/tasky.